### ISSUE TITLE
Past Covariates for DirectTabular and RecursiveTabular models

### ISSUE THREAD

**Issue by chrissype** [awslabs/autogluon#5429]
## Description
The DirectTabular and RecursiveTabular models don't support past covariates. I like these models and they appear to work well in practice.

Based on my understanding of how the models work, it sounds like it should be relatively easy to include this information, by adding items and their lags to the (presumably large) input matrix that gets generated. 

Is my understanding correct, or is there a reason this hasn't been implemented?

## References
- [docs](https://auto.gluon.ai/stable/tutorials/timeseries/forecasting-model-zoo.html#additional-features)

**Comment by shchur** [comment]
Hi @chrissype, we investigated adding the value of the past covariates lagged by `prediction_length` to `DirectTabular` / `RecursiveTabular` models. On a benchmark of ~30 datasets with past covariates this led to worse forecast accuracy in ~70% of the cases, so I decided not to merge this into mainline. It is possible that the past covariates were not very informative in these cases.

I guess we could add this as an optional hyperparameter to the models, but I don't think I will have time to merge this by v1.5.

**Comment by chrissype** [comment]
Interesting data point thank you, although I guess I also think that implies that the covariates weren't useful.

In a classic case of the XY problem, this is really because I have noted that increasing the amount of missing data in the context window at inference time generally leads quantile predictions to collapse, when really they should be expanding to indicate much reduced certainty.

I believe this is because the autogluon data pipeline imputes mean values or ffills constant values for missing data, making it look like the context has a constant value as you increase the amount of data that's missing, so the models naturally predict constant values with high certainty. In particular the tabular models should natively support nan values and allow one to do this.

So I wanted to add an indicator variable for whether the value is missing. But obviously this only makes sense as a past covariate.

**Comment by shchur** [comment]
Hi @chrissype, I would recommend trying the [PerStepTabularModel](https://auto.gluon.ai/stable/tutorials/timeseries/forecasting-model-zoo.html#autogluon.timeseries.models.PerStepTabularModel) and setting `eval_metric` to "WQL" or "SQL" for quantile mode. 

The NaN values in the original series are already ignored during fitting for the tabular models, so I wouldn't expect the indicator to have a major effect. You can still try providing it as a known covariate and just setting the value in the forecast horizon to all ones. I don't think that using the lagged value as a past covariate would be helpful here.

**Comment by chrissype** [comment]
I saw that in the release notes and thought it looked interesting so perhaps I'll try it. Naively though I would expect that would give better quantile calibration with respect to increasing horizon, but not with respect to missing context data. For my use case I'm also targeting long horizon predictions (~280 steps), 7 different quantiles, and a large dataset so i suspect train and inference might get too slow but I can give it a try.

Also, is there a reason not to include nans in the training pipeline? As i understand it, most tree models including lightgbm natively handle nans and allow splits on that value.

And furthermore, is my intuition about train time imputed values collapsing the quantiles correct? If so it sounds from your description like there might be a drift between train time and inference time for the tree models - at train time the nans are ignored, whereas at inference time they are imputed. The only way i want the missing data to be useful is in indicating to the models (in general, I'm not only using the tabular ones) that the predictions should be less confident.
